samples/net/dsa/src/dsa_lldp.c

```diff
@@ -0,0 +1,185 @@
+		  struct eth_addr *origin_addr)
+{
+	int ret, len;
+	char buffer[DSA_BUF_SIZ], s[24];
+	struct sockaddr_ll dst;
+	struct ethernet_context *ctx = net_if_l2_data(iface);
+	struct net_eth_hdr *eth_hdr = (struct net_eth_hdr *) buffer;
+	uint8_t *p = &buffer[sizeof(struct net_eth_hdr)];
+	uint8_t *pb = p;
+
+	lan = ctx->dsa_port_idx;
+
+	dst.sll_ifindex = net_if_get_by_iface(iface);
+	/* Construct the Ethernet header */
+	memset(buffer, 0, DSA_BUF_SIZ);
+	/* Ethernet header */
+	/* Take MAC address assigned to LAN port */
+	memcpy(eth_hdr->src.addr, net_if_get_link_addr(iface)->addr, ETH_ALEN);
+	eth_hdr->dst.addr[0] = MCAST_DEST_MAC0;
+	eth_hdr->dst.addr[1] = MCAST_DEST_MAC1;
+	eth_hdr->dst.addr[2] = MCAST_DEST_MAC2;
+	eth_hdr->dst.addr[3] = MCAST_DEST_MAC3;
+	eth_hdr->dst.addr[4] = MCAST_DEST_MAC4;
+	eth_hdr->dst.addr[5] = MCAST_DEST_MAC5;
+
+	/* Ethertype field */
+	eth_hdr->type = htons(LLDP_ETHER_TYPE);
+
+	/* LLDP packet data */
+	/* Chasis ID */
+	dsa_buf_write_be16((LLDP_TLV_CHASSIS_ID << 9) | (ETH_ALEN + 1), &p);
+	*p++ = 4; /* subtype */
+	memcpy(p, net_if_get_link_addr(iface)->addr, ETH_ALEN);
+	p += ETH_ALEN;
+
+	/* PORT ID */
+	dsa_buf_write_be16((LLDP_TLV_PORT_ID << 9) | (ETH_ALEN + 1), &p);
+	*p++ = 3; /* subtype */
+	memcpy(p, net_if_get_link_addr(iface)->addr, ETH_ALEN);
+	p += ETH_ALEN;
+
+	/* TTL ID */
+	dsa_buf_write_be16((LLDP_TLV_TTL << 9) | 2, &p);
+	*p++ = 0; /* TTL field is 2 bytes long */
+	*p++ = 120;
+
+	/* SYSTEM NAME */
+	memset(s, 0, sizeof(s));
+	sprintf(s, "ip_k66f LAN:%d", lan);
+	dsa_buf_write_be16((LLDP_TLV_SYSTEM_NAME << 9) | strlen(s), &p);
+	memcpy(p, s, strlen(s));
+	p += strlen(s);
+
+	len = sizeof(struct net_eth_hdr) + (p - pb);
+	ret = sendto(pd->sock, buffer, len, 0, (const struct sockaddr *)&dst,
+		     sizeof(struct sockaddr_ll));
+	if (ret < 0) {
+		LOG_ERR("Failed to send, errno %d", errno);
+	}
+
+	return 0;
+}
+
+void dsa_lldp_print_info(uint8_t *lldp_p, uint8_t lanid)
+{
+	uint16_t tl, length;
+	uint8_t type, subtype;
+	uint8_t *p, t1, t2;
+	char *t;
+
+	LOG_INF("LLDP pkt recv -> lan%d", lanid);
+	do {
+		/* In-buffer data is stored as big endian */
+		t1 = *lldp_p++;
+		t2 = *lldp_p++;
+		tl = (uint16_t) t1 << 8 | t2;
+
+		/* Get type and length */
+		type = tl >> 9;
+		length = tl & 0x1FF;
+
+		switch (type) {
+		case LLDP_TLV_CHASSIS_ID:
+		case LLDP_TLV_PORT_ID:
+			/* Extract subtype */
+			subtype = *lldp_p++;
+			length--;
+			break;
+		}
+
+		p = lldp_p;
+
+		switch (type) {
+		case LLDP_TLV_END_LLDPDU:
+			return;
+		case LLDP_TLV_CHASSIS_ID:
+			LOG_INF("\tCHASIS ID:\t%02x:%02x:%02x:%02x:%02x:%02x",
+				p[0], p[1], p[2], p[3], p[4], p[5]);
+			break;
+		case LLDP_TLV_PORT_ID:
+			LOG_INF("\tPORT ID:\t%02x:%02x:%02x:%02x:%02x:%02x",
+				p[0], p[1], p[2], p[3], p[4], p[5]);
+			break;
+		case LLDP_TLV_TTL:
+			/* TTL field has 2 bytes in BE */
+			LOG_INF("\tTTL:\t\t%ds", (uint16_t) p[0] << 8 | p[1]);
+			break;
+		case LLDP_TLV_SYSTEM_NAME:
+			t = malloc(length+1);
```

Comment:
Looks like an overkill to use malloc() here. You could use a buffer with some sane length instead.

Reply:
Here I would also prefer to keep `malloc`. I do know that it is the overkill at least in this example, but any "real" use case  (and hence code) would use `malloc`.

Reply:
> but any "real" use case (and hence code) would use malloc.

Many (majority of ?) embedded programmers would not use malloc. Moreover, some embedded programming standards (e.g. MISRA C) explicitly prohibit use of dynamic memory allocation. The more malloc's are in the Zephyr's codebase (even if in samples/), the stranger look Zephyr's aspirations to adhere to embedded programming best practices. Bottom line: unless the malloc is really due (e.g., there're samples for the POSIX subsystem, and malloc() is unalienable part of the POSIX), then not using malloc() is more "future-proof" (regarding where the Zephyr project may want to go).


My comment about that line: should be spaces around `+` in `length+1`.

Reply:
Ok. I will remove `malloc()` and replace it with the buffer automatically allocated.